refactor(compiler): Add `getPotentialTemplateDirectives` API method.

dylhunn opened the pull request:
`getPotentialTemplateDirectives` returns possible directives which can be used in the provided context, whether already in scope or requiring an import.

This is necessary to implement auto-import support for standalone components in the language service.

This is fourth in an ongoing sequence of PRs, including #47181, #47180, and #47166.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:
```diff
@@ -716,14 +708,42 @@ export class TemplateTypeCheckerImpl implements TemplateTypeChecker {
   }
+
+  private scopeDataOfDirectiveMeta(
+      typeChecker: ts.TypeChecker, dep: DirectiveMeta, isInScope: boolean): DirectiveInScope|null {
```
Don't pass `isInScope` through the interface here.

This function doesn't care about whether or not `dep` is in scope. Passing an argument to a function only to reflect it back out into the return value is a code smell.

What this is telling us is that the return type here is wrong. There is a concept of "directive data that consumers of `TemplateTypeChecker` care about" which is produced from `DirectiveMeta`.  Additionally, consumers care whether the directive is in scope or not, but _that_ part of the information is not produced from `DirectiveMeta` and relates to how we found the `DirectiveMeta` in the first place.

So we have 2 options:

1. Give both types names:

```typescript
interface DirectiveTypeCheckerData {
  isComponent: boolean;
  isStructural: boolean;
  // etc.
}

interface PotentialDirective /*DirectiveInScope*/ extends DirectiveTypeCheckerData {
  isInScope: boolean;
}
```

2. Create the base type dynamically.

```typescript
private scopeDataOfDirectiveMeta(...): Omit<PotentialDirective, 'isInScope'>|null {
  ...
}
```

dylhunn replied:
I've chosen the second option, since this is just factoring out common code, and we're not expecting to use the type elsewhere.

dylhunn commented on packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:
```diff
@@ -549,12 +550,22 @@ export class TemplateTypeCheckerImpl implements TemplateTypeChecker {
+    const inScopeDirectives = this.getScopeData(component)?.directives ?? [];
+    const resultingDirectives = new Map<ClassDeclaration<DeclarationNode>, DirectiveInScope>();
+    // First, all in scope directives can be used.
+    inScopeDirectives.forEach(d => resultingDirectives.set(d.ref.node, d));
```
Done, but why the objection to `forEach`?

alxhub replied:
The debugging experience of `for..of` is _significantly_ better :)

dylhunn committed:
refactor(compiler): Rename `DirectiveInScope` -> `PotentialDirective`

After implementing `getPotentialTemplateDirectives`, we will use this data struture to represent both in-scope and out-of-scope directives. So this rename is an advance cleanup.

dylhunn commented on packages/compiler-cli/src/ngtsc/scope/src/global.ts:
```diff
@@ -0,0 +1,42 @@
+import {DirectiveMeta, MetadataReader, MetadataRegistry, MetaKind, NgModuleMeta, PipeMeta} from '../../metadata';
+import {ClassDeclaration} from '../../reflection';
+
+export class GlobalDirectiveRegistry implements MetadataRegistry, MetadataReader {
```
I'm not 100% sure 'Global' is really the right name either.

For now, the differences with `LocalMetadataRegistry` are slight -- we're allowing access to the set of all known directives, rather than looking up one specific `DirectiveMeta`. My belief is that we'll want to extend this registry in the near future to support things like automatic imports from `.d.ts` files, and more intelligent import specifier tracking.

alxhub replied:
> My belief is that we'll want to extend this registry in the near future to support things like automatic imports from .d.ts files

We probably wouldn't put these two behaviors into the same registry, but would have separate classes for the separate logic.

> and more intelligent import specifier tracking.

Until we get to this point, I would prefer to avoid creating a new class and just use `LocalMetadataRegistry`.

dylhunn replied:
Done. In order to delete `GlobalTraitRegistry`, I needed to extend `LocalMetadataRegistry` with the `getKnownDirectives()` functionality. I thought it would be cleanest to also create a new interface
```
export interface MetadataReaderWithIndex extends MetadataReader {
  getKnownDirectives(): Set<ClassDeclaration>;
}
```
That way we have the correct type for a `MetadataReader` that also supports this new functionality.

alxhub commented on packages/compiler-cli/src/ngtsc/metadata/src/api.ts:
```diff
@@ -250,6 +250,13 @@ export interface MetadataReader {
+ * A MetadataReader which also allows access to the set of all known directive classes.
+ */
+export interface MetadataReaderWithIndex extends MetadataReader {
+  getKnownDirectives(): Set<ClassDeclaration>;
```
```suggestion
  getKnownDirectives(): ReadonlySet<ClassDeclaration>;
```

alxhub commented on packages/compiler-cli/src/ngtsc/metadata/src/registry.ts:
```diff
@@ -40,6 +40,10 @@ export class LocalMetadataRegistry implements MetadataRegistry, MetadataReader {
     this.pipes.set(meta.ref.node, meta);
   }
+
+  getKnownDirectives(): Set<ClassDeclaration> {
```
Alternatively, we could return the iterable type directly, or maintain a separate `Set`. We should avoid re-allocating this on every request though.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/checker.ts:
```diff
@@ -886,7 +886,7 @@ class SingleShimTypeCheckingHost extends SingleFileTypeCheckingHost {
 interface ScopeData {
-  directives: DirectiveInScope[];
+  directives: PotentialDirective[];
   pipes: PipeInScope[];
```
Should we not do the same with `PipeInScope` -> `PotentialPipe`?

dylhunn committed:
refactor(compiler): Add `getPotentialTemplateDirectives` API method.

`getPotentialTemplateDirectives` returns possible directives which can be used in the provided context, whether already in scope or requiring an import.

This is necessary to implement auto-import support for standalone components in the language service.

dylhunn committed:
refactor(compiler): Rename `PipeInScope` -> `PotentialPipe`

Now that `getPotentialTemplateDirectives` uses `PotentialDirective`, we should rename `PipeInScope` to match.

AndrewKushnir commented:
This PR was merged into the repository by commit 29194638c11fe642b97987e9118ec9fa061ae9d4.
